applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp

```diff
@@ -165,30 +165,32 @@ void DistanceModificationProcess::ModifyDistance() {
     }
     // Case in where the original distance needs to be kept to track the interface (e.g. FSI)
     else {
-        const unsigned int NumThreads = OpenMPUtils::GetNumThreads();
-        std::vector<std::vector<unsigned int>> AuxModifiedDistancesIDs(NumThreads);
-        std::vector<std::vector<double>> AuxModifiedDistancesValues(NumThreads);
 
-        #pragma omp parallel shared(AuxModifiedDistancesIDs, AuxModifiedDistancesValues)
+        const unsigned int num_chunks = 2 * OpenMPUtils::GetNumThreads();
+        OpenMPUtils::PartitionVector partition_vec;
+        OpenMPUtils::DivideInPartitions(r_nodes.size(),num_chunks,partition_vec);
+
+        #pragma omp parallel for
+        for (unsigned int i_chunk = 0; i_chunk < num_chunks; ++i_chunk)
         {
-            const int ThreadId = OpenMPUtils::ThisThread();             // Get the thread id
-            std::vector<unsigned int>   LocalModifiedDistancesIDs;      // Local modified distances nodes id vector
-            std::vector<double>      LocalModifiedDistancesValues;      // Local modified distances original values vector
-
-            #pragma omp for
-            for (int k = 0; k < static_cast<int>(rNodes.size()); ++k) {
-                ModelPart::NodesContainerType::iterator itNode = rNodes.begin() + k;
-                const double h = itNode->FastGetSolutionStepValue(NODAL_H);
+            auto nodes_begin = r_nodes.begin() + partition_vec[i_chunk];
+            auto nodes_end = r_nodes.begin() + partition_vec[i_chunk + 1];
+
+            for (auto it_node = nodes_begin; it_node != nodes_end; ++it_node) {
+                const double h = it_node->FastGetSolutionStepValue(NODAL_H);
                 const double tol_d = mDistanceThreshold*h;
-                double &d = itNode->FastGetSolutionStepValue(DISTANCE);
+                double &d = it_node->FastGetSolutionStepValue(DISTANCE);
 
                 // Check if the distance values are close to zero
                 // If proceeds, set the tolerance as distance value
                 if(std::abs(d) < tol_d){
 
                     // Store the original distance to be recovered at the end of the step
-                    LocalModifiedDistancesIDs.push_back(itNode->Id());
-                    LocalModifiedDistancesValues.push_back(d);
+                    #pragma omp critical
```

Comment:
i wouldn't make one critical per each entry, otherwise the cost is enormous. create an auxilary local array and push back to it without critical, then within the critical add the auxiliary array to the global one "at once"